File: python/leetcodepy/wildcard_matching.py
"""
Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).

The matching should cover the entire input string (not partial).

Note:

s could be empty and contains only lowercase letters a-z.
p could be empty and contains only lowercase letters a-z, and characters like ? or *.

Example 1:

Input:
s = "aa"
p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".

Example 2:

Input:
s = "aa"
p = "*"
Output: true
Explanation: '*' matches any sequence.

Example 3:

Input:
s = "cb"
p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.

Example 4:

Input:
s = "adceb"
p = "*a*b"
Output: true
Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".

Example 5:

Input:
s = "acdcb"
p = "a*c?b"
Output: false
"""


class Solution1:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        size_s, size_p = len(s), len(p)
        dp = [[False] * (size_p + 1) for _ in range(size_s + 1)]
        dp[0][0] = True
        for j in range(1, size_p + 1):
            if p[j - 1] == '*':
                dp[0][j] = True
            else:
                break
        for i in range(1, size_s + 1):
            for j in range(1, size_p + 1):
                if p[j - 1] == '?' or p[j - 1] == s[i - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                elif p[j - 1] == '*':
                    dp[i][j] = dp[i][j - 1] or dp[i - 1][j]
        return dp[size_s][size_p]


class Solution2:
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        len_s, len_p = len(s), len(p)
        i, j = 0, 0
        star, last_str = -1, -1
        while i < len_s:
            ch = p[j] if j < len_p else None
            if ch == '?' or ch == s[i]:
                i += 1
                j += 1
            elif ch == '*':
                last_str = i
                star = j
                j += 1
            elif star != -1:
                last_str += 1
                i = last_str
                j = star + 1
            else:
                return False
        while j < len_p and p[j] == '*':
            j += 1
        return j == len_p

File: java/src/main/java/io/binac/leetcode/WildcardMatching.java
package io.binac.leetcode;

/**
 * <p>Given an input string (<code>s</code>) and a pattern (<code>p</code>), implement wildcard pattern matching with support for <code>'?'</code> and <code>'*'</code> where:</p>
 *
 * <ul>
 * 	<li><code>'?'</code> Matches any single character.</li>
 * 	<li><code>'*'</code> Matches any sequence of characters (including the empty sequence).</li>
 * </ul>
 *
 * <p>The matching should cover the <strong>entire</strong> input string (not partial).</p>
 *
 * <p>&nbsp;</p>
 * <p><strong>Example 1:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "aa", p = "a"
 * <strong>Output:</strong> false
 * <strong>Explanation:</strong> "a" does not match the entire string "aa".
 * </pre>
 *
 * <p><strong>Example 2:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "aa", p = "*"
 * <strong>Output:</strong> true
 * <strong>Explanation:</strong>&nbsp;'*' matches any sequence.
 * </pre>
 *
 * <p><strong>Example 3:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "cb", p = "?a"
 * <strong>Output:</strong> false
 * <strong>Explanation:</strong>&nbsp;'?' matches 'c', but the second letter is 'a', which does not match 'b'.
 * </pre>
 *
 * <p><strong>Example 4:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "adceb", p = "*a*b"
 * <strong>Output:</strong> true
 * <strong>Explanation:</strong>&nbsp;The first '*' matches the empty sequence, while the second '*' matches the substring "dce".
 * </pre>
 *
 * <p><strong>Example 5:</strong></p>
 *
 * <pre><strong>Input:</strong> s = "acdcb", p = "a*c?b"
 * <strong>Output:</strong> false
 * </pre>
 *
 * <p>&nbsp;</p>
 * <p><strong>Constraints:</strong></p>
 *
 * <ul>
 * 	<li><code>0 &lt;= s.length, p.length &lt;= 2000</code></li>
 * 	<li><code>s</code> contains only lowercase English letters.</li>
 * 	<li><code>p</code> contains only lowercase English letters, <code>'?'</code> or <code>'*'</code>.</li>
 * </ul>
 */
public class WildcardMatching {
    public static class Solution1 {
        public boolean isMatch(String s, String p) {
            final int m = s.length(), n = p.length();
            boolean[][] match = new boolean[m + 1][n + 1];
            match[0][0] = true;
            for (int j = 1; j <= n && p.charAt(j - 1) == '*'; ++j)
                match[0][j] = true;

            char ch;
            for (int i = 1; i <= m; ++i) {
                for (int j = 1; j <= n; ++j) {
                    ch = p.charAt(j - 1);
                    if (ch == '*')
                        match[i][j] = match[i - 1][j] || match[i][j - 1];
                    else if (ch == '?' || ch == s.charAt(i - 1))
                        match[i][j] = match[i - 1][j - 1];
                }
            }
            return match[m][n];
        }
    }

    public static class Solution2 {
        public boolean isMatch(String s, String p) {
            final int m = s.length(), n = p.length();
            int i = 0, j = 0;
            int star = -1, lastI = -1;
            char ch;
            while (i < m) {
                ch = j < n ? p.charAt(j) : '\0';
                if (ch == '?' || ch == s.charAt(i)) {
                    ++i;
                    ++j;
                } else if (ch == '*') {
                    lastI = i;
                    star = j;
                    ++j;
                } else if (star != -1) {
                    i = ++lastI;
                    j = star + 1;
                } else {
                    return false;
                }
            }
            while (j < n && p.charAt(j) == '*') ++j;
            return j == n;
        }
    }
}
